Core/Scripts: Core support for Heroic Leap

Boomper opened the pull request:
- after using Heroic Leap player will now actually jump and do AOE damage after landing
- also adds support for Glyph of Heroic Leap (adds 70% movement boost after landing from heroic leap) and Improved Heroic Leap (dreanor perk, resets cooldown of Taunt after landing from heroic leap)

Closes: https://github.com/TrinityCore/TrinityCore/issues/14134
Partialy based upon (the CheckCast part): https://github.com/TrinityCore/TrinityCore/issues/9313

Apologies if I've misstyped somewhere, english is not my native tongue.
P.S.: Happy jumping around Azeroth :-)

Boomper committed:
Core/Scripts: Core support for Heroic Leap

 - after using Heroic Leap player will now actually jump and do AOE damage after landing
 - also adds support for Glyph of Heroic Leap (adds 70% movement boost after landing from heroic leap) and Improved Heroic Leap (dreanor perk, resets cooldown of Taunt after landing from heroic leap)

Frytiks commented:
Looks cool +1

tkrokli commented on sql/updates/world/spell_script_name_heroic_leap.sql:
```diff
@@ -0,0 +1,4 @@
+DELETE FROM `spell_script_names` WHERE `spell_id` IN ('6544', '178368');
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+('6544','spell_warr_heroic_leap'),
+('178368','spell_warr_heroic_leap_jump');
```
Add a newline/linefeed here. While you are on it, rename the file to something like 2015_XX_XX_NN_world.sql

Shauren commented on src/server/game/Spells/Spell.cpp:
```diff
@@ -4727,7 +4727,7 @@ SpellCastResult Spell::CheckCast(bool strict)
         }
 
-        if (!m_caster->GetSpellHistory()->IsReady(m_spellInfo))
+        if (!m_caster->GetSpellHistory()->IsReady(m_spellInfo) && !(_triggeredCastFlags & TRIGGERED_IGNORE_SPELL_AND_CATEGORY_CD))
```
the fuck

Boomper replied:
@Shauren: It made sense at the time since..., I'll make it work without this.

Boomper replied:
Can't decide what is lesser evil, either removing the cooldown of spell 178368 before casting the spell or removing it from the SpellCategory cooldowns. Do you have any suggestions or a third less hacky way how to make it work?

Shauren commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -948,6 +953,100 @@ class spell_warr_vigilance_trigger : public SpellScriptLoader
+            {
+                if (dest->GetPositionZ() > GetCaster()->GetPositionZ() + 4.0f)
+                    return SPELL_FAILED_NOPATH;
+                else if (GetCaster()->HasAuraType(SPELL_AURA_MOD_ROOT))
```
This should check root movement flag, not aura (as there can be other things applying the root, stun for example)

robinsch replied:
But those should not result in SPELL_FAILED_ROOTED. SPELL_FAILED_ROOTED is only for auras with SPELL_AURA_MOD_ROOT. 

Client doesn't say "Can't be done while rooted" when having stun auras.

Shauren commented on src/server/scripts/Spells/spell_warrior.cpp:
```diff
@@ -948,6 +953,100 @@ class spell_warr_vigilance_trigger : public SpellScriptLoader
+        {
+            if (WorldLocation const* dest = GetExplTargetDest())
+            {
+                if (dest->GetPositionZ() > GetCaster()->GetPositionZ() + 4.0f)
```
And this probably should use the same mechanic as charge - check if path can be generated instead of such a simple elevation comparison

Boomper replied:
![wowscrnshot_053115_152404](https://cloud.githubusercontent.com/assets/799249/7902073/b3999a58-07a9-11e5-821e-f34fc0e7ed0e.jpg)
![wowscrnshot_053115_152411](https://cloud.githubusercontent.com/assets/799249/7902074/b3a06234-07a9-11e5-8c55-ce316f3937a4.jpg)

As far as I know it doesn't use the pathfinding system, from playing around with it on retail it suggests that it uses height comparsion and the z value is around 4yards.

Shauren commented:
It behaves differently inside instances (more restricted, requires proper path)

Boomper commented:
So if the map is Instanceable it should use the CheckCast for spells like charge?

Shauren commented:
I meant the path check

Boomper committed:
Core/Scripts: Heroic Leap

 - added newline at the end of sql file, and rename of the sql file
 - checking for unitmovement flag instead of hasauratype
 - if the map is instanceable it does the charge path check
 - and removed the thingy that Shauren didn't liked in Spell.cpp (replaced with ResetCooldown)

robinsch commented:
Why do you check twice for the path?

Boomper commented:
I've copy-pasted (and adjusted) the CheckCast part from here:
https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Spells/Spell.cpp#L5161
the first looks only for straight path and the second one looks for any path.

robinsch commented:
wops, my bad. Now I get it, thanks.

Robens commented:
why don't you add a checkcast for all spells with effect SPELL_EFFECT_JUMP_DEST in spell::checkcast? instead of creating a hook

Boomper commented:
Because the initial spell has dummy effect

Robens commented:
You can return spelljump->CheckCast(false);  less code repetition, fewer errors

Boomper commented:
First time hearing/reading about that method, must've been hidding under some pretty big rocks...

But I'm still not going to put it in there because I have no idea if the same rules apply for all spells with SPELL_EFFECT_JUMP_DEST (there's quite a lot of them)

tkrokli commented:
Just a friendly tip (even if it is nice to see that you changed the SQL filename to the current standard) :
"2015_05_31_00_world.sql" has already been committed, so you need to use a filename that has not been used before / will not be used in the near future. From [[CONTRIBUTING.md]](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/CONTRIBUTING.md) in the 3.3.5 source :
`Since it's very unlikely that your Pull Request will be merged on the day that you open it, please name the files with an impossible date to avoid merging issues ie: 2015_13_32_00_world.sql`

Boomper commented:
On branch 6.x the last sql update for world DB is https://github.com/TrinityCore/TrinityCore/blob/6.x/sql/updates/world/2015_05_29_00_world.sql , I think that you might be looking at wrong branch (3.3.5)

tkrokli commented:
Yes, you're right. Sorry about that. I will delete my comments about this.

jackpoz commented:
This PR has conflicts that must be resolved, please rebase it on a more recent commit.
Is there any additional feedback or can it be merged after the conflicts have been resolved ?

ShinDarth commented:
if conflicts are the only reason of not merging this PR, I suggest to merge it with commit (fixing manually the conflicts) since last comment of @jackpoz was 23 days ago and the PR author doesn't seem to be there

tkrokli commented:
If https://github.com/TrinityCore/TrinityCore/pull/14849#issuecomment-156689633 is meant to be in general, the author does not have time to work on this any more.
